django/utils/datastructures.py

```diff
@@ -280,3 +281,84 @@ def __getitem__(self, key):
         if use_func:
             return self.func(value)
         return value
+
+
+def _destruct_iterable_mapping_values(data):
+    for i, elem in enumerate(data):
+        if len(elem) != 2:
+            raise ValueError("dictionary update sequence element #{} has "
+                             "length {}; 2 is required".format(i, len(elem)))
+
+        if not isinstance(elem[0], str):
+            raise ValueError('Element key invalid, only strings are allowed')
+
+        yield tuple(elem)
+
+
+class ImmutableCaseInsensitiveDict(Mapping):
```

Comment:
This isn't actually Immutable. I didn't trace far back enough to see if the WSGI `environ` is immutable but cursory inspection would indicate that `.META` is not actually immutable although it would probably be a bad idea to manipulate it (I might be wrong on this point).

Reply:
@santiagobasulto What are your thoughts on renaming this to `CaseInsensitiveMapping`? `Mapping` being the stdlib ABC you're implementing. (c.f. `ImmutableList.complain()`: we're taking measures there to avoid mutation. I don't think we want or need to get into any of that here.)